### PR TITLE
[#246] Autosave tasks, once valid

### PR DESCRIPTION
**Pull request by tonythomas01**
TODO:
* Would look more good if the text is moved to the right end.

### CODE REVIEW

File: web/js/tasks.js [hunk 1]
```diff
@@ -884,6 +884,37 @@ Ext.onReady(function(){
...
+            var valids = true;
+            for(var panel=0; panel<panels.getCount(); panel++) {
+                console.log();
+                if (!panels.get(panel).initTimeField.isValid() ||
+                    !panels.get(panel).endTimeField.isValid() ||
+                    !panels.get(panel).customerComboBox.getRawValue() ||
+                    !panels.get(panel).projectComboBox.getRawValue() ||
+                    !panels.get(panel).storyField.getRawValue() ||
+                    !panels.get(panel).descriptionTextArea.getRawValue()
+                ) {
```
Comment (jaragunde) [inline web/js/tasks.js]: The only compulsory fields are init and end time. Tasks may be saved with empty customer, project, etc.

**Comment by jaragunde** [comment]
Have you tried to set `autosave: true` in the tasks Store? Maybe it works...

**Comment by tonythomas01** [comment]
`autosave:true` makes it save on every change (even submitting when the data is not entered, throwing up an app error ).

File: web/js/tasks.js (hunk 1 above)
Comment (jaragunde) [inline web/js/tasks.js]: Hmmm, it's better not to duplicate the validation code. It already lives in the `saveTasks()` function. Maybe you could split that method in two parts (validateTasks and saveTasks) to make it easier to reuse.

File: web/js/tasks.js [hunk 2]
```diff
@@ -884,6 +884,37 @@ Ext.onReady(function(){
...
+            }
+            if (valids) {
+                myStore.each(function(r) {
+                    if (r.data['story'] != undefined)
+                        r.data['story'] = xmlencode(r.data['story']);
+                    if (r.data['text'] != undefined)
+                        r.data['text'] = xmlencode(r.data['text']);
+                });
+                myStore.save();
+                unsavedChanges = false;
```
Comment (jaragunde) [inline web/js/tasks.js]: Don't duplicate this code either, better rework `saveTasks()` a little bit to make it work as you want so you may call it from here.

Also, `unsavedChanges = false` should not happen here, only after the store is successfully saved.

File: web/js/tasks.js [hunk 3]
```diff
@@ -884,6 +884,37 @@ Ext.onReady(function(){
...
+                Ext.getCmp('status_display').setText("Status: Auto saved at "+ new Date());
```
Comment (jaragunde) [inline web/js/tasks.js]: It's not really important if the tasks were manually or automatically saved. You may move this message to the save event in the Store object.

File: web/js/tasks.js [hunk 4]
```diff
@@ -882,34 +883,53 @@ Ext.onReady(function(){
...
+        // First we check if the time fields of all records are valid, and then save
+        if (validateTasks()) {
+            addToMyStore();
         } else  // Otherwise, we print the error message
           App.setAlert(false, "Check For Invalid Field Values");
     }
 
+    // Implement autosave of data, when valid contents are typed in to the task fields
+    window.setInterval(function () {
+        if( unsavedChanges == true ) {
```
Comment (jaragunde) [inline web/js/tasks.js]: Don't compare a boolean like this, you could have written `if(unsavedChanges)....`. In this case, you may even use `if(isUnsaved())`

**Comment by jaragunde** [comment]
Hi Tony, this is looking great now. I've just noticed that the *status* indicator does not notice when you have made changes to a task. For example, create a task, save it and then modify its description. 

Another situation is when you open a date that has some saved tasks, the indicator says "status: draft" even when you haven't done any changes yet. Maybe it should say "unchanged".

I think you could implement the status indicator with a callback that triggers periodically and checks the status of `isUnsaved()`. This function already takes into account changes inside the tasks. Maybe you could reuse the same function that triggers the save operation.

**Comment by jaragunde** [comment]
I noticed an error in `isUnsaved()` implementation when testing your patch. Please rebase to the latest master :)

**Comment by jaragunde** [comment]
> I think you could implement the status indicator with a callback that triggers periodically and checks the status of isUnsaved().

Or you could modify the value of _status_ in the `update`, `remove` events of the store.

**Pull request merged**